### DISCUSSION TITLE
recommended way to input and store parameters?

### DISCUSSION THREAD

**Discussion by petrelharp** [MesserLab/SLiM#214] [Ideas]
I often want to specify a bunch of parameters in each script, which are set to defaults unless I change them, either on the command line or in a simple file. I used to do it this way:
```
 if (!exists("SIGMA"))
    defineConstant("SIGMA", 0.1);
```
but (a) this is a bit awkward if there's a lot of them, and more importantly (b) I'd also like to save all the values of these parameters in the tree sequence metadata when I write it out; currently to do that I'd need to build a dictionary with all the parameters.

**Comment by petrelharp** [comment]
Here's how I've started doing it:
```
initialize() {
    initializeSLiMModelType("nonWF");
    initializeMutationRate(0.0);
    initializeMutationType("m1", 0.5, "f", 0.0);
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 100-1);
    initializeRecombinationRate(1e-8);

    defaults = Dictionary(
        "NAME", "abc",
        "SIGMA", 2
    );
    for (k in defaults.allKeys) {
        if (!exists(k)) {
            defineConstant(k, defaults.getValue(k));
        } else {
            defaults.setValue(k, executeLambda(paste(c(k, ";"), sep='')));
        }
    }
    defineGlobal("params", defaults);
}

1 early() {    catn(params);  }
```
So, in my script I can just use SIGMA because it's a constant; at the end I'll pass `params` to `sim.treeSeqOutput('out.trees', metadata=params)`. If I run just
```
slim test.slim
```
it prints out
```
{SIGMA=2;NAME="abc";}
```
while if I do
```
slim -d SIGMA=5 -d "NAME='newname'" test.slim
```
I get
```
{NAME="newname";SIGMA=5;}
```

**Reply by petrelharp** [reply]
The next thing I want to do is to have parameters in a file that gets read in. If I make a file `params.slim` and in it put
```
Dictionary(
  "NAME", "xyz",
  "SIGMA", 5
)
```
and I insert the code
```
	if (fileExists("params.slim")) {
		input = executeLambda(paste(c(readFile("params.slim"), ";"), sep=" "));
		defaults.addKeysAndValuesFrom(input);
	}
```
in the above script, then that'll let me be able to specify the parameters in a config file. But it would be a bit nicer if the config file was of the form
```
NAME="xyz";
SIGMA=5;
```
... probably there's a nice way to set that up?

**Reply by grahamgower** [reply]
One alternative is to read the params in from a json file. Something like:
```
    lines = readFile(json_file);
    if (isNULL(lines)) {
        stop(json_file + ": failed to load model");
    }
    string = paste(lines, sep="\n");
    dict = Dictionary(string);
```

The disadvantage is you can't annotate the json file with comments.

**Reply by petrelharp** [reply]
Oh, right! I forgot that you could make Dictionaries from a JSON string! But that is too bad about the lack of comments, hmph.

**Reply by petrelharp** [reply]
I guess a format like this could work:
```
{
    "X": {
          "value": 2,
          "description": "the number 2"
    },
    "name": {
           "value": "xyz",
          "description": "used for filename output"
    }
}
```

**Reply by petrelharp** [reply]
... combined with
```
	if (fileExists("params.slim")) {
		input = Dictionary(paste(readFile("params.slim"), sep="\n"));
                for (k in input.allKeys)
                       defaults.setValue(k, input.getValue(k).getValue("value"));
	}
```

**Reply by grahamgower** [reply]
Hmm, I think it would be nicer to have a wrapper function which strips out (possibly indented) lines starting with `//` (or maybe `#`?), after calling `readFile()`. Only supporting comments that are on a line of their own ought to be easy enough (and really isn't a huge burden for users).

PS: I reckon `readFile()` could fail for reasons other than the file not existing. E.g. lack of permissions, or an I/O issue.
